src/pyclass.rs

```diff
@@ -154,122 +190,197 @@ unsafe fn create_type_object_impl(
+                ffi::structmember::PyMemberDef {
+                    name: name.as_ptr() as _,
+                    type_code: ffi::structmember::T_PYSSIZET,
+                    offset,
+                    flags: ffi::structmember::READONLY,
+                    doc: std::ptr::null_mut(),
+                }
+            }
+
+            let mut members = Vec::new();
+
+            // __dict__ support
+            if let Some(dict_offset) = dict_offset {
+                members.push(offset_def("__dictoffset__\0", dict_offset));
+            }
+
+            // weakref support
+            if let Some(weaklist_offset) = weaklist_offset {
+                members.push(offset_def("__weaklistoffset__\0", weaklist_offset));
+            }
+
+            // Safety: Py_tp_members expects a raw vec of PyMemberDef
+            unsafe { self.push_raw_vec(ffi::Py_tp_members, members) };
+        }
+
+        // Setting buffer protocols, tp_dictoffset and tp_weaklistoffset via slots doesn't work until
+        // Python 3.9, so on older versions we must manually fixup the type object.
+        #[cfg(all(not(Py_LIMITED_API), not(Py_3_9)))]
+        {
+            self.cleanup
+                .push(Box::new(move |builder, type_object| unsafe {
+                    (*(*type_object).tp_as_buffer).bf_getbuffer = builder.buffer_procs.bf_getbuffer;
+                    (*(*type_object).tp_as_buffer).bf_releasebuffer =
+                        builder.buffer_procs.bf_releasebuffer;
+
+                    if let Some(dict_offset) = dict_offset {
+                        (*type_object).tp_dictoffset = dict_offset;
+                    }
+
+                    if let Some(weaklist_offset) = weaklist_offset {
+                        (*type_object).tp_weaklistoffset = weaklist_offset;
+                    }
+                }));
         }
+        self
     }
 
-    // Setting buffer protocols, tp_dictoffset and tp_weaklistoffset via slots doesn't work until
-    // Python 3.9, so on older versions we must manually fixup the type object.
-    #[cfg(not(Py_3_9))]
-    {
-        (*(*_type_object).tp_as_buffer).bf_getbuffer = buffer_procs.bf_getbuffer;
-        (*(*_type_object).tp_as_buffer).bf_releasebuffer = buffer_procs.bf_releasebuffer;
+    fn build(
+        mut self,
+        py: Python<'_>,
+        name: &'static str,
+        module_name: Option<&'static str>,
+        basicsize: usize,
+    ) -> PyResult<*mut ffi::PyTypeObject> {
+        self.finalize_methods_and_properties();
 
-        if let Some(dict_offset) = dict_offset {
-            (*_type_object).tp_dictoffset = dict_offset;
+        if !self.has_new {
+            // Safety: This is the correct slot type for Py_tp_new
+            unsafe { self.push_slot(ffi::Py_tp_new, no_constructor_defined as *mut c_void) }
         }
 
-        if let Some(weaklist_offset) = weaklist_offset {
-            (*_type_object).tp_weaklistoffset = weaklist_offset;
+        if !self.has_dealloc {
```

Comment:
Just a thought: in principle the typestate pattern could move this (and most other checks in this file) from runtime to compile time. I don't think it'd be worth it though.

Reply:
Given that this code is expected to only execute once per pyclass I agree that runtime checks are preferred here over getting the compiler to do lots of codegen / compile-time checks.